File: study/2022/06-W2/wg.shin.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        h, n = haystack, needle
        N = len(h)
        M = len(n)
        if M == 0: return 0
        
        for i in range(N-M+1):
            offset = i
            j = 0
            while j < M:
                if h[offset+j] != n[j]:
                    break
                j += 1
            if j == M:
                return i
            
        return -1
# Runtime: 44 ms, faster than 45.83% of Python3 online submissions for Implement strStr().
# Memory Usage: 14 MB, less than 15.29% of Python3 online submissions for Implement strStr().

File: study/2022/10-W2/wg.shin.dart
class Solution {
  int findJudge(int n, List<List<int>> trust) {
    Map<int, List<Set<int>>> graph = Map.fromIterable(Iterable<int>.generate(n, (i) => i + 1), value: (i) => [{}, {}]);

    for (List<int> edge in trust) {
      graph[edge[1]]![0].add(edge[0]);
      graph[edge[0]]![1].add(edge[1]);
    }

    for (int key in graph.keys) {
      if (graph[key]![0].length == n - 1 && graph[key]![1].length == 0) {
        return key;
      }
    }

    return -1;
  }
}

/*
Runtime: 386 ms, faster than 100% of Dart online submissions for Find the Town Judge.
Memory Usage: 187.8 MB, less than 33.33% of Dart online submissions for Find the Town Judge.
 */
